File: autobox/docker/manager.py
import asyncio
import logging
import platform
from typing import Any, Dict, List, Optional

import docker
from docker.errors import NotFound

logger = logging.getLogger(__name__)


class DockerManager:
    def __init__(self):
        try:
            self.client = docker.from_env()
            self.image_name = "autobox-engine:latest"
        except Exception as e:
            logger.error(f"Failed to initialize Docker client: {e}")
            self.client = None

    async def list_running_simulations(self) -> List[Dict[str, str]]:
        if not self.client:
            return []

        try:
            containers = self.client.containers.list(
                filters={"label": "com.autobox.simulation=true"}
            )
            return [
                {
                    "id": container.short_id,
                    "name": container.name,
                    "status": container.status,
                    "created": container.attrs["Created"],
                }
                for container in containers
            ]
        except Exception as e:
            logger.error(f"Error listing containers: {e}")
            return []

    async def start_simulation(
        self, config_path: str, metrics_path: str, server_config_path: str = None
    ) -> Optional[str]:
        if not self.client:
            raise RuntimeError("Docker client not initialized")

        try:
            import json
            import os
            from pathlib import Path

            config_name = Path(config_path).stem

            if not server_config_path:
                server_config_path = str(Path.home() / ".autobox/config/server.json")

            api_port = 9000
            try:
                with open(server_config_path, "r") as f:
                    server_config = json.load(f)
                    api_port = server_config.get("port", 9000)
                    logger.info(f"Using API port {api_port} from server config")
            except Exception as e:
                logger.warning(
                    f"Could not read server config, using default port 9000: {e}"
                )

            import socket

            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.bind(("", 0))
                host_port = s.getsockname()[1]
            logger.info(f"Allocated host port {host_port} for container")

            running_in_docker = (
                os.path.exists("/.dockerenv") or str(Path.home()) == "/root"
            )

            container_port = f"{api_port}/tcp"

            if running_in_docker:
                host_home = os.environ.get("HOST_HOME")

                if not host_home:
                    username = os.environ.get("HOST_USER", "root")

                    if platform.system() == "Darwin" or os.path.exists("/Users"):
                        host_home = f"/Users/{username}"
                    else:
                        host_home = f"/home/{username}"

                host_autobox_path = os.path.expanduser(f"{host_home}/.autobox")

                container = self.client.containers.run(
                    self.image_name,
                    command=[
                        "--config",
                        f"/app/configs/simulations/{config_path.split('/')[-1]}",
                        "--metrics",
                        f"/app/configs/metrics/{metrics_path.split('/')[-1]}",
                        "--server",
                        f"/app/configs/server/{server_config_path.split('/')[-1]}"
                        if server_config_path
                        else "",
                    ],
                    detach=True,
                    environment={
                        "OPENAI_API_KEY": os.environ.get("OPENAI_API_KEY", ""),
                        "OBJC_DISABLE_INITIALIZE_FORK_SAFETY": "TRUE",
                        "PYTHONUNBUFFERED": "1",
                        "AUTOBOX_EXTERNAL_PORT": str(host_port),
                    },
                    volumes={
                        f"{host_autobox_path}/config/simulations": {
                            "bind": "/app/configs/simulations",
                            "mode": "ro",
                        },
                        f"{host_autobox_path}/config/metrics": {
                            "bind": "/app/configs/metrics",
                            "mode": "ro",
                        },
                        f"{host_autobox_path}/config": {
                            "bind": "/app/configs/server",
                            "mode": "ro",
                        },
                    },
                    ports={container_port: host_port},
                    name=f"autobox-sim-{asyncio.get_event_loop().time():.0f}",
                    labels={
                        "autobox.api_port": str(api_port),
                        "com.autobox.simulation": "true",
                        "com.autobox.name": config_name,
                        "com.autobox.config_path": str(config_path),
                        "com.autobox.created_at": str(asyncio.get_event_loop().time()),
                    },
                    remove=False,
                )
            else:
                container = self.client.containers.run(
                    self.image_name,
                    command=[
                        "--config",
                        f"/app/configs/simulations/{config_path.split('/')[-1]}",
                        "--metrics",
                        f"/app/configs/metrics/{metrics_path.split('/')[-1]}",
                        "--server",
                        f"/app/configs/server/{server_config_path.split('/')[-1]}",
                    ],
                    detach=True,
                    environment={
                        "OPENAI_API_KEY": os.environ.get("OPENAI_API_KEY", ""),
                        "OBJC_DISABLE_INITIALIZE_FORK_SAFETY": "TRUE",
                        "PYTHONUNBUFFERED": "1",
                        "AUTOBOX_EXTERNAL_PORT": str(host_port),
                    },
                    volumes={
                        str(Path.home() / ".autobox/config/simulations"): {
                            "bind": "/app/configs/simulations",
                            "mode": "ro",
                        },
                        str(Path.home() / ".autobox/config/metrics"): {
                            "bind": "/app/configs/metrics",
                            "mode": "ro",
                        },
                        str(Path.home() / ".autobox/config"): {
                            "bind": "/app/configs/server",
                            "mode": "ro",
                        },
                    },
                    ports={container_port: host_port},
                    name=f"autobox-sim-{asyncio.get_event_loop().time():.0f}",
                    labels={
                        "com.autobox.simulation": "true",
                        "com.autobox.name": config_name,
                        "com.autobox.config_path": str(config_path),
                        "com.autobox.created_at": str(asyncio.get_event_loop().time()),
                    },
                    remove=False,
                )

            container.reload()
            port_info = container.attrs.get("NetworkSettings", {}).get("Ports", {})
            if container_port in port_info and port_info[container_port]:
                host_port = port_info[container_port][0]["HostPort"]
                logger.info(
                    f"Started simulation container: {container.short_id} with API on port {host_port}"
                )
            else:
                logger.info(f"Started simulation container: {container.short_id}")

            return container.short_id

        except Exception as e:
            logger.error(f"Error starting simulation: {e}")
            raise

    async def stop_simulation(self, container_id: str) -> bool:
        if not self.client:
            return False

        try:
            container = self.client.containers.get(container_id)
            container.stop(timeout=10)
            container.remove()
            logger.info(f"Stopped simulation container: {container_id}")
            return True
        except NotFound:
            logger.warning(f"Container not found: {container_id}")
            return False
        except Exception as e:
            logger.error(f"Error stopping container: {e}")
            return False

    async def stop_all_simulations(self) -> Dict[str, Any]:
        if not self.client:
            return {"error": "Docker client not available", "stopped": [], "failed": []}

        stopped = []
        failed = []

        try:
            containers = self.client.containers.list(
                filters={"label": "com.autobox.simulation=true", "status": "running"}
            )

            for container in containers:
                container_id = container.id[:12]
                try:
                    container.stop(timeout=10)
                    container.remove()
                    stopped.append({"id": container_id, "name": container.name})
                    logger.info(f"Stopped simulation container: {container.name}")
                except Exception as e:
                    failed.append(
                        {"id": container_id, "name": container.name, "error": str(e)}
                    )
                    logger.error(f"Failed to stop container {container.name}: {e}")

            return {
                "stopped": stopped,
                "failed": failed,
                "total_stopped": len(stopped),
                "total_failed": len(failed),
            }

        except Exception as e:
            logger.error(f"Error listing/stopping containers: {e}")
            return {
                "error": f"Failed to stop all simulations: {str(e)}",
                "stopped": stopped,
                "failed": failed,
            }

    async def get_container_status(self, container_id: str) -> Optional[Dict]:
        if not self.client:
            return None

        try:
            container = self.client.containers.get(container_id)
            return {
                "id": container.short_id,
                "name": container.name,
                "status": container.status,
                "running": container.status == "running",
            }
        except NotFound:
            return None
        except Exception as e:
            logger.error(f"Error getting container status: {e}")
            return None

    async def get_logs(self, container_id: str, tail: int = 100) -> Optional[str]:
        if not self.client:
            return None

        try:
            container = self.client.containers.get(container_id)
            logs = container.logs(tail=tail, timestamps=True).decode("utf-8")
            return logs
        except NotFound:
            return None
        except Exception as e:
            logger.error(f"Error getting logs: {e}")
            return None

    async def get_container_stats(self, container_id: str) -> Optional[Dict]:
        if not self.client:
            return None

        try:
            container = self.client.containers.get(container_id)
            stats = container.stats(stream=False)

            cpu_delta = (
                stats["cpu_stats"]["cpu_usage"]["total_usage"]
                - stats["precpu_stats"]["cpu_usage"]["total_usage"]
            )
            system_delta = (
                stats["cpu_stats"]["system_cpu_usage"]
                - stats["precpu_stats"]["system_cpu_usage"]
            )
            cpu_percent = 0.0
            if system_delta > 0 and cpu_delta > 0:
                cpu_percent = (cpu_delta / system_delta) * 100.0

            memory_usage = stats["memory_stats"].get("usage", 0)
            memory_limit = stats["memory_stats"].get("limit", 0)
            memory_percent = (
                (memory_usage / memory_limit * 100) if memory_limit > 0 else 0
            )

            network_rx = 0
            network_tx = 0
            if "networks" in stats:
                for interface in stats["networks"].values():
                    network_rx += interface.get("rx_bytes", 0)
                    network_tx += interface.get("tx_bytes", 0)

            return {
                "cpu_percent": round(cpu_percent, 2),
                "memory_usage_mb": round(memory_usage / 1024 / 1024, 2),
                "memory_limit_mb": round(memory_limit / 1024 / 1024, 2),
                "memory_percent": round(memory_percent, 2),
                "network_rx_mb": round(network_rx / 1024 / 1024, 2),
                "network_tx_mb": round(network_tx / 1024 / 1024, 2),
            }
        except NotFound:
            return None
        except Exception as e:
            logger.error(f"Error getting container stats: {e}")
            return None

    async def get_simulation_api_status(
        self, container_id: str, port: int = None
    ) -> Optional[Dict]:
        if not self.client:
            return None

        try:
            container = self.client.containers.get(container_id)

            if container.status != "running":
                return None

            if port is None:
                port = int(container.labels.get("autobox.api_port", "9000"))

            container_info = container.attrs
            ports = container_info.get("NetworkSettings", {}).get("Ports", {})

            port_key = f"{port}/tcp"
            if port_key in ports and ports[port_key]:
                host_port = ports[port_key][0]["HostPort"]
                host_ip = ports[port_key][0].get("HostIp", "localhost")
                if host_ip == "0.0.0.0":
                    host_ip = "localhost"

                import httpx

                async with httpx.AsyncClient() as client:
                    try:
                        response = await client.get(
                            f"http://{host_ip}:{host_port}/status",
                            timeout=5.0,
                        )
                        if response.status_code == 200:
                            return response.json()
                    except Exception as e:
                        logger.warning(f"Failed to get status from API: {e}")

            networks = container_info.get("NetworkSettings", {}).get("Networks", {})
            for network_name, network_info in networks.items():
                ip_address = network_info.get("IPAddress")
                if ip_address:
                    import httpx

                    async with httpx.AsyncClient() as client:
                        try:
                            response = await client.get(
                                f"http://{ip_address}:{port}/status",
                                timeout=5.0,
                            )
                            if response.status_code == 200:
                                return response.json()
                        except Exception:
                            pass

            return None
        except Exception as e:
            logger.error(f"Error getting API status: {e}")
            return None

    async def get_simulation_api_metrics(
        self, container_id: str, port: int = None
    ) -> Optional[Dict]:
        if not self.client:
            return None

        try:
            container = self.client.containers.get(container_id)

            if container.status != "running":
                return None

            if port is None:
                port = int(container.labels.get("autobox.api_port", "9000"))

            container_info = container.attrs
            ports = container_info.get("NetworkSettings", {}).get("Ports", {})

            port_key = f"{port}/tcp"
            if port_key in ports and ports[port_key]:
                host_port = ports[port_key][0]["HostPort"]
                host_ip = ports[port_key][0].get("HostIp", "localhost")
                if host_ip == "0.0.0.0":
                    host_ip = "localhost"

                import httpx

                async with httpx.AsyncClient() as client:
                    try:
                        response = await client.get(
                            f"http://{host_ip}:{host_port}/metrics",
                            timeout=5.0,
                        )
                        if response.status_code == 200:
                            return response.json()
                    except Exception as e:
                        logger.warning(f"Failed to get metrics from API: {e}")

            networks = container_info.get("NetworkSettings", {}).get("Networks", {})
            for network_name, network_info in networks.items():
                ip_address = network_info.get("IPAddress")
                if ip_address:
                    import httpx

                    async with httpx.AsyncClient() as client:
                        try:
                            response = await client.get(
                                f"http://{ip_address}:{port}/metrics",
                                timeout=5.0,
                            )
                            if response.status_code == 200:
                                return response.json()
                        except Exception:
                            pass

            return None
        except Exception as e:
            logger.error(f"Error getting API metrics: {e}")
            return None

File: src/docker/manager.ts
import Docker from 'dockerode';
import { homedir, platform } from 'os';
import { join } from 'path';
import type { ContainerStats, SimulationMetrics, SimulationStatus } from '../types/index.js';
import { logger } from '../utils/logger.js';

interface StartSimulationOptions {
  configPath: string;
  metricsPath: string;
  serverConfigPath?: string;
}

export class DockerManager {
  private client: Docker;
  private imageName = 'autobox-engine:latest';

  constructor() {
    try {
      this.client = new Docker();
    } catch (error) {
      logger.error('Failed to initialize Docker client:', error);
      throw error;
    }
  }

  async listRunningSimulations(): Promise<SimulationStatus[]> {
    try {
      const containers = await this.client.listContainers({
        filters: { label: ['com.autobox.simulation=true'] },
      });

      return containers.map((container) => ({
        id: container.Id.substring(0, 12),
        name: container.Names[0]?.replace(/^\//, '') || container.Id.substring(0, 12),
        status: container.State,
        created: new Date(container.Created * 1000).toISOString(),
        running: container.State === 'running',
      }));
    } catch (error) {
      logger.error('Error listing containers:', error);
      return [];
    }
  }

  async startSimulation(options: StartSimulationOptions): Promise<string> {
    const { configPath, metricsPath, serverConfigPath } = options;
    const configName = configPath.split('/').pop()?.replace('.json', '') || 'unknown';

    const serverPath = serverConfigPath || join(homedir(), '.autobox/config/server.json');

    const apiPort = 9000;
    const hostPort = await this.findAvailablePort();

    logger.info(`Starting simulation ${configName} on host port ${hostPort}`);

    const homeDir = homedir();
    const autoboxPath = join(homeDir, '.autobox');
    const runningInDocker = process.env.HOST_HOME !== undefined;

    const hostAutoboxPath = runningInDocker
      ? this.resolveHostPath(process.env.HOST_HOME || homeDir)
      : autoboxPath;

    const volumes = {
      [`${hostAutoboxPath}/config/simulations`]: {
        bind: '/app/configs/simulations',
        mode: 'ro',
      },
      [`${hostAutoboxPath}/config/metrics`]: {
        bind: '/app/configs/metrics',
        mode: 'ro',
      },
      [`${hostAutoboxPath}/config`]: {
        bind: '/app/configs/server',
        mode: 'ro',
      },
    };

    const containerName = `autobox-sim-${Date.now()}`;

    const container = await this.client.createContainer({
      Image: this.imageName,
      name: containerName,
      Cmd: [
        '--config',
        `/app/configs/simulations/${configPath.split('/').pop()}`,
        '--metrics',
        `/app/configs/metrics/${metricsPath.split('/').pop()}`,
        '--server',
        `/app/configs/server/${serverPath.split('/').pop()}`,
      ],
      Env: [
        `OPENAI_API_KEY=${process.env.OPENAI_API_KEY || ''}`,
        'OBJC_DISABLE_INITIALIZE_FORK_SAFETY=TRUE',
        'PYTHONUNBUFFERED=1',
        `AUTOBOX_EXTERNAL_PORT=${hostPort}`,
      ],
      HostConfig: {
        Binds: Object.entries(volumes).map(([src, { bind, mode }]) => `${src}:${bind}:${mode}`),
        PortBindings: {
          [`${apiPort}/tcp`]: [{ HostPort: hostPort.toString() }],
        },
        AutoRemove: false,
      },
      Labels: {
        'com.autobox.simulation': 'true',
        'com.autobox.name': configName,
        'com.autobox.config_path': configPath,
        'com.autobox.created_at': Date.now().toString(),
        'autobox.api_port': apiPort.toString(),
      },
    });

    await container.start();
    logger.info(`Started simulation container: ${container.id.substring(0, 12)}`);

    return container.id.substring(0, 12);
  }

  async stopSimulation(containerId: string): Promise<boolean> {
    try {
      const container = this.client.getContainer(containerId);
      await container.stop({ t: 10 });
      await container.remove();
      logger.info(`Stopped simulation container: ${containerId}`);
      return true;
    } catch (error) {
      if ((error as { statusCode?: number }).statusCode === 404) {
        logger.warn(`Container not found: ${containerId}`);
        return false;
      }
      logger.error('Error stopping container:', error);
      return false;
    }
  }

  async stopAllSimulations(): Promise<{
    stopped: Array<{ id: string; name: string }>;
    failed: Array<{ id: string; name: string; error: string }>;
    total_stopped: number;
    total_failed: number;
  }> {
    const stopped: Array<{ id: string; name: string }> = [];
    const failed: Array<{ id: string; name: string; error: string }> = [];

    try {
      const containers = await this.client.listContainers({
        filters: {
          label: ['com.autobox.simulation=true'],
          status: ['running'],
        },
      });

      await Promise.all(
        containers.map(async (containerInfo) => {
          const containerId = containerInfo.Id.substring(0, 12);
          const containerName =
            containerInfo.Names[0]?.replace(/^\//, '') || containerInfo.Id.substring(0, 12);

          try {
            const container = this.client.getContainer(containerInfo.Id);
            await container.stop({ t: 10 });
            await container.remove();
            stopped.push({ id: containerId, name: containerName });
            logger.info(`Stopped simulation container: ${containerName}`);
          } catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            failed.push({ id: containerId, name: containerName, error: errorMessage });
            logger.error(`Failed to stop container ${containerName}:`, error);
          }
        })
      );

      return {
        stopped,
        failed,
        total_stopped: stopped.length,
        total_failed: failed.length,
      };
    } catch (error) {
      logger.error('Error listing/stopping containers:', error);
      return { stopped, failed, total_stopped: 0, total_failed: 0 };
    }
  }

  async getContainerStatus(
    containerId: string
  ): Promise<{ id: string; name: string; status: string; running: boolean } | null> {
    try {
      const container = this.client.getContainer(containerId);
      const info = await container.inspect();

      return {
        id: info.Id.substring(0, 12),
        name: info.Name.replace(/^\//, ''),
        status: info.State.Status,
        running: info.State.Running,
      };
    } catch (error) {
      if ((error as { statusCode?: number }).statusCode === 404) {
        return null;
      }
      logger.error('Error getting container status:', error);
      return null;
    }
  }

  async getLogs(containerId: string, tail = 100): Promise<string | null> {
    try {
      const container = this.client.getContainer(containerId);
      const logs = await container.logs({
        stdout: true,
        stderr: true,
        tail,
        timestamps: true,
      });

      return logs.toString('utf-8');
    } catch (error) {
      if ((error as { statusCode?: number }).statusCode === 404) {
        return null;
      }
      logger.error('Error getting logs:', error);
      return null;
    }
  }

  async getContainerStats(containerId: string): Promise<ContainerStats | null> {
    try {
      const container = this.client.getContainer(containerId);
      const stats = await container.stats({ stream: false });

      const cpuDelta =
        stats.cpu_stats.cpu_usage.total_usage - stats.precpu_stats.cpu_usage.total_usage;
      const systemDelta = stats.cpu_stats.system_cpu_usage - stats.precpu_stats.system_cpu_usage;
      const cpuPercent = systemDelta > 0 && cpuDelta > 0 ? (cpuDelta / systemDelta) * 100 : 0;

      const memoryUsage = stats.memory_stats.usage || 0;
      const memoryLimit = stats.memory_stats.limit || 0;
      const memoryPercent = memoryLimit > 0 ? (memoryUsage / memoryLimit) * 100 : 0;

      let networkRx = 0;
      let networkTx = 0;
      if (stats.networks) {
        Object.values(stats.networks).forEach((network: any) => {
          networkRx += network.rx_bytes || 0;
          networkTx += network.tx_bytes || 0;
        });
      }

      return {
        cpu_percent: Math.round(cpuPercent * 100) / 100,
        memory_usage_mb: Math.round((memoryUsage / 1024 / 1024) * 100) / 100,
        memory_limit_mb: Math.round((memoryLimit / 1024 / 1024) * 100) / 100,
        memory_percent: Math.round(memoryPercent * 100) / 100,
        network_rx_mb: Math.round((networkRx / 1024 / 1024) * 100) / 100,
        network_tx_mb: Math.round((networkTx / 1024 / 1024) * 100) / 100,
      };
    } catch (error) {
      if ((error as { statusCode?: number }).statusCode === 404) {
        return null;
      }
      logger.error('Error getting container stats:', error);
      return null;
    }
  }

  async getSimulationApiMetrics(
    containerId: string,
    includeDockerStats = true
  ): Promise<SimulationMetrics | null> {
    try {
      const container = this.client.getContainer(containerId);
      const info = await container.inspect();

      if (!info.State.Running) {
        return null;
      }

      const apiPort = info.Config.Labels?.['autobox.api_port'] || '9000';
      const portBindings = info.NetworkSettings.Ports?.[`${apiPort}/tcp`];

      let metrics: SimulationMetrics | null = null;

      if (portBindings && portBindings.length > 0) {
        const hostPort = portBindings[0].HostPort;
        const hostIp = portBindings[0].HostIp === '0.0.0.0' ? 'localhost' : portBindings[0].HostIp;

        metrics = await this.fetchMetricsFromApi(`http://${hostIp}:${hostPort}/metrics`);
      }

      if (!metrics) {
        const networks = info.NetworkSettings.Networks || {};
        for (const network of Object.values(networks)) {
          if (network.IPAddress) {
            metrics = await this.fetchMetricsFromApi(
              `http://${network.IPAddress}:${apiPort}/metrics`
            );
            if (metrics) break;
          }
        }
      }

      if (!metrics) {
        return null;
      }

      metrics.simulation_id = containerId;
      metrics.timestamp = Date.now();

      if (includeDockerStats) {
        const dockerStats = await this.getContainerStats(containerId);
        if (dockerStats) {
          metrics.docker_stats = dockerStats;
        }
      }

      return metrics;
    } catch (error) {
      logger.error('Error getting API metrics:', error);
      return null;
    }
  }

  async instructAgent(
    containerId: string,
    agentName: string,
    instruction: string
  ): Promise<{ success: boolean; message?: string; error?: string }> {
    try {
      const container = this.client.getContainer(containerId);
      const info = await container.inspect();

      if (!info.State.Running) {
        return { success: false, error: `Simulation ${containerId} is not running` };
      }

      const apiPort = info.Config.Labels?.['autobox.api_port'] || '9000';
      const portBindings = info.NetworkSettings.Ports?.[`${apiPort}/tcp`];

      if (portBindings && portBindings.length > 0) {
        const hostPort = portBindings[0].HostPort;
        const hostIp = portBindings[0].HostIp === '0.0.0.0' ? 'localhost' : portBindings[0].HostIp;
        const url = `http://${hostIp}:${hostPort}/instructions/agents/${agentName.toLowerCase()}`;

        return await this.sendInstructionToApi(url, instruction);
      }

      const networks = info.NetworkSettings.Networks || {};
      for (const network of Object.values(networks)) {
        if (network.IPAddress) {
          const url = `http://${network.IPAddress}:${apiPort}/instructions/agents/${agentName.toLowerCase()}`;
          const result = await this.sendInstructionToApi(url, instruction);
          if (result.success) return result;
        }
      }

      return { success: false, error: 'Could not connect to simulation API' };
    } catch (error) {
      logger.error('Error instructing agent:', error);
      return { success: false, error: error instanceof Error ? error.message : String(error) };
    }
  }

  private async fetchMetricsFromApi(url: string): Promise<SimulationMetrics | null> {
    try {
      const response = await fetch(url, {
        signal: AbortSignal.timeout(5000),
      });

      if (response.ok) {
        return (await response.json()) as SimulationMetrics;
      }
    } catch (error) {
      logger.debug(`Failed to fetch metrics from ${url}:`, error);
    }
    return null;
  }

  private async sendInstructionToApi(
    url: string,
    instruction: string
  ): Promise<{ success: boolean; message?: string; error?: string }> {
    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ instruction }),
        signal: AbortSignal.timeout(10000),
      });

      if (response.ok) {
        return {
          success: true,
          message: 'Instruction sent successfully',
        };
      }

      return {
        success: false,
        error: `API returned status ${response.status}`,
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  private async findAvailablePort(): Promise<number> {
    const net = await import('net');
    return new Promise((resolve, reject) => {
      const server = net.createServer();
      server.unref();
      server.on('error', reject);
      server.listen(0, () => {
        const address = server.address();
        if (address && typeof address !== 'string') {
          const port = address.port;
          server.close(() => resolve(port));
        } else {
          reject(new Error('Failed to get port'));
        }
      });
    });
  }

  private resolveHostPath(hostHome: string): string {
    const username = process.env.HOST_USER || 'root';

    if (platform() === 'darwin' || hostHome.includes('/Users')) {
      return `/Users/${username}/.autobox`;
    }

    return `/home/${username}/.autobox`;
  }
}
